### PR TITLE
C API: support pinnable get

### PR DESCRIPTION
**Pull request by siddontang**

### CODE REVIEW

File: db/c.cc [hunk 1]
```diff
@@ -3088,6 +3092,52 @@ void rocksdb_delete_file_in_range_cf(
 
 void rocksdb_free(void* ptr) { free(ptr); }
 
+rocksdb_pinnableslice_t* rocksdb_pin_get(rocksdb_t* db,
```
Comment (IslamAbdelRahman) [inline db/c.cc]: maybe a better name will be `rocksdb_get_pinned`

File: db/c.cc [hunk 2]
```diff
@@ -3088,6 +3092,52 @@ void rocksdb_delete_file_in_range_cf(
...
+      SaveError(errptr, s);
+    }
+    return NULL;
+  }
+  return v;
+}
+
+void rocksdb_pinnableslice_destroy(rocksdb_pinnableslice_t* v) { delete v; }
+
+const char* rocksdb_pinnableslice_value(const rocksdb_pinnableslice_t* v,
```
Comment (IslamAbdelRahman) [inline db/c.cc]: should these 2 functions be in the c_test file ?

File: HISTORY.md [hunk 3]
```diff
@@ -1,7 +1,8 @@
 # Rocksdb Change Log
 ## Unreleased
 ### Public API Change
-* Introduce WriteBatch::PopSavePoint to pop the most recent save point explicitly
+* Introduce WriteBatch::PopSavePoint to pop the most recent save point explicitly.
+* Introduce C rocksdb_pin_get/rocksdb_pin_get_cf API to get value with PinnableSlice.
```
Comment (IslamAbdelRahman) [inline HISTORY.md]: We generally don't add entries for C / Java API changes

File: db/c.cc (hunk 2 above)
Comment (siddontang) [inline db/c.cc]: yes

File: HISTORY.md (hunk 3 above)
Comment (siddontang) [inline HISTORY.md]: I will remove it.

File: db/c.cc (hunk 1 above)
Comment (siddontang) [inline db/c.cc]: +1

**Comment by siddontang** [comment]
PTAL @IslamAbdelRahman

File: db/c_test.c [hunk 4]
```diff
@@ -126,6 +126,32 @@ static void CheckGetCF(
   Free(&val);
 }
 
+static void CheckPinGet(rocksdb_t* db, const rocksdb_readoptions_t* options,
+                        const char* key, const char* expected) {
+  char* err = NULL;
+  size_t val_len;
+  const char* val;
+  rocksdb_pinnableslice_t* p;
```
Comment (maysamyabandeh) [inline db/c_test.c]: Why does it have to be created on heap? Can't it be simply an on-stack object that is passed to rocksdb_get_pinned?

Comment (siddontang) [inline db/c_test.c]: so we need a new function rocksdb_create_pinnableslice to create the object like other C API do(e,g rocksdb_writebatch_create), I think it is not necessary here.

Btw, I think it is better to keep the same way with rocksdb_get, return the value and free it manually outside.

Comment (maysamyabandeh) [inline db/c_test.c]: Are not we concerned of the cost of creating the object on heap?

Comment (siddontang) [inline db/c_test.c]: I don't think this is a big problem. 

Btw, rocksdb_pinnableslice_t is just a typedef, the compiler may throw error:

```
error: variable has incomplete type 'rocksdb_pinnableslice_t' (aka 'struct rocksdb_pinnableslice_t')
  rocksdb_pinnableslice_t pin;
```

Comment (siddontang) [inline db/c_test.c]: I don't think this is a big problem. 

Btw, rocksdb_pinnableslice_t is just a typedef, the compiler may throw error:

```
error: variable has incomplete type 'rocksdb_pinnableslice_t' (aka 'struct rocksdb_pinnableslice_t')
  rocksdb_pinnableslice_t pin;
```

Comment (maysamyabandeh) [inline db/c_test.c]: @IslamAbdelRahman What do you think?

**Comment by yiwu-arbug** [comment]
@IslamAbdelRahman Any comment?

**Comment by IslamAbdelRahman** [comment]
LGTM but let's move
```
rocksdb_pinnableslice_destroy()
rocksdb_pinnableslice_value()
```
to c_test before landing

**Comment by siddontang** [comment]
Hi @IslamAbdelRahman 

I don't know why to move rocksdb_pinnableslice_destroy, rocksdb_pinnableslice_value to c_test?
Outer will also use these functions, not only in the test.

**Pull request closed without merging**